### ISSUE TITLE
pnpm update not updating local package

### ISSUE THREAD

**Issue by aegorx** [pnpm/pnpm#5372]
### pnpm version:

7.12.0

### Code to reproduce the issue:

install local pkg

1. pnpm add xxx

"xxx": "file:xxx",

2. pnpm update xxx

### Expected behavior:

update pkg xxx

### Additional information:

 - `node -v` prints:  node 16
 - Windows, macOS, or Linux?: macOS

**Comment by kbsali** [comment]
I am having the same problem

**Comment by sundjly** [comment]
I also  have the same problem;  node 16 Linux

**Comment by Rayologist** [comment]
I use docker on node 16 bullseye slim and encountered the same problem

**Comment by itsdevcoffee** [comment]
Same deal is happening to me. Fedora 36 Linux. pnpm@7.12.1

**Comment by mjgerace** [comment]
This must be a new bug... I have some decent info for the maintainers of the project. I spent most of my night binary-eliminating packages, and found that for some reason, pnpm has a bug where it won't actually complete installation upon `pnpm install` or `pnpm update`, specifically for me when installing one of the following 4 packages:

```
    "@storybook/addon-essentials": "6.5.10",
    "@storybook/addon-links": "6.5.10",
    "@storybook/builder-vite": "0.2.0",
    "@storybook/react": "6.5.10",
```

This muse be a new bug, as I literally found pnpm completely unusable - it would quietly fail (even with the json reporter turned on) in the install process. I'd never get a postinstall message or any package +++ or --- which is how I picked up on this. Please prioritize this bug @zkochan , willing to help in any way I can. No peer deps on those in any capacity, fyi.

**Comment by Rayologist** [comment]
I have the same issue when I use docker for nodejs, whose version is node:16-bullseye-slim. 
The problem is that when I install bcrypt@5.0.1 use pnpm, it won't build the necessary binary file of bcrypt.
The current solution is that 
```
RUN [ ! -f ./node_modules/bcrypt/lib/binding/napi-v3/bcrypt_lib.node ] && cd node_modules/bcrypt && make
```

**Comment by zgriesinger** [comment]
I'm having the exact same issue as @mjgerace 

It's happening _only_ on 7.11 and not on 7.12

Will try to get a repro

**Comment by mjgerace** [comment]
> I cannot reproduce this. Please try to create exact steps to reproduce. Maybe with a sample repo.

pnpm version 7.12

Hi, here is an information dump:

I am in a monorepo with 15 packages or so. In 3 of these packages, I have the below specified in my package.json. I have removed them all to get a working lockfile. if I add _any_ of the following to even 1 package and `pnpm install` (from the root folder of the monorepo, set up with turbo 1.4.7), I don't actually get to the end of `pnpm install`:

```
"@storybook/addon-essentials": "6.5.10",
"@storybook/addon-links": "6.5.10",
"@storybook/builder-vite": "0.2.0",
"@storybook/react": "6.5.10",
```

Here is an example of a `pnpm install` without those included anywhere ( a working `pnpm install`, in this case we see it do nothing since this has just succeeded - but it would work even if we added a "safe" package, IE one not causing issues):
```
pnpm install
Scope: all 15 workspace projects
Lockfile is up to date, resolution step is skipped
Already up to date
. postinstall$ pnpm patch rc-dock@3.2.10
│ You can now edit the following folder: /private/var/folders/ks/h83ybgfx28dg2wrrjh95zb8h0000gn/T/d9a29d04c34f4a76293eedc00812fab7
└─ Done in 417ms
```

Here is an example of including `@storybook/react` in one of the packages:

```
pnpm install
Scope: all 15 workspace projects
packageName1                                      |  WARN  deprecated blah
packageName2                                      |  WARN  deprecated blah
packageName3                            |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName5                                    |  WARN  deprecated blah
packageName5                                    |  WARN  deprecated blah
packageName6                                      |  WARN  deprecated blah
packageName2                                      |  WARN  deprecated blah
packageName6                                      |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName4                                   |  WARN  deprecated blah
packageName7                                    |  WARN  deprecated blah
packageName7                                    |  WARN  deprecated blah
Progress: resolved 2767, reused 2693, downloaded 1, added 0
{email-here}@michaels-mbp Terminal %
```

As you can see, yarn install doesn't succeed once pnpm tries to get this package. I have tried clearing the store as well just to remove cache. The 4 dependencies that are problematic are not listed in any other dependency (IE, peer, dev, regular) at the moment.

**Comment by zkochan** [comment]
@mjgerace I was about to reproduce it. Here's a repro repo: https://github.com/zkochan/pnpm-issue-5372

**Comment by zkochan** [comment]
Looks like some deadlock. Though I am not sure why it fails silently. Narrowed down to this dependency graph:

![image](https://user-images.githubusercontent.com/1927579/191843081-7ae0b738-7552-4ada-839f-6c9d165f4636.png)

**Comment by mjgerace** [comment]
@zkochan 

I do have auto-install-peers=true, I'd expect that this commit contains the issue since many people report 7.11 functional (I cannot confirm, have not tried): https://github.com/pnpm/pnpm/commit/8cecfcbe3f511ceb914b7ee1a316ac391212d339

**Comment by zkochan** [comment]
Yes, I think that is true. I am trying to fix this.

**Comment by zkochan** [comment]
fixed in v7.12.2
